src/storage/src/render/persist_sink.rs

```diff
@@ -1183,27 +1187,56 @@ where
 
                 let result = {
                     let _permit = busy_signal.acquire().await;
-                    write.compare_and_append_batch(
-                        &mut to_append[..],
-                        batch_lower.clone(),
-                        batch_upper.clone(),
-                    )
-                    .await
-                    .expect("Invalid usage")
+
+                    if *read_only_rx.borrow() {
+                        // We synthesize an `UpperMismatch` so that we can go
+                        // through the same logic below for trimming down our
+                        // batches.
+                        //
+                        // Notably, we are not trying to be smart, and teach the
+                        // write operator about read-only mode. Writing down
+                        // those batches does not append anything to the persist
+                        // shard, and it would be a hassle to figure out in the
+                        // write workers how to trim down batches in read-only
+                        // mode, when the shard upper advances.
+                        //
+                        // Right here, in the logic below, we have all we need
+                        // for figuring out how to trim our batches.
+                        let current_upper = write.fetch_recent_upper().await.clone();
```

Comment:
I think there is an edge case here if the frontier of the desired collection has already advanced to the empty frontier (meaning no more events will be generated from the input handle) but we don't act on it because we are in read-only mode. If the read/write replica is shut down before it manages to itself commit the desired collection until the empty frontier will this operator wake up when the read-only flag flips to false? I don't see any await on `read_only_rx` which makes me think this relies on a steady stream of input updates for getting unstuck